File: Backtracking/combinationSum2.java
class Solution {
    HashSet<List<Integer>> res = new HashSet();

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        backTrack(candidates, target, new ArrayList(), 0);
        return new ArrayList(res);
    }

    public void backTrack(int[] candidates, int target, ArrayList<Integer> list, int rejected) {
        if (target < 0)
            return;
        if (target == 0) {
            ArrayList l = new ArrayList(list);
            // Collections.sort(l);
            res.add(l);
            return;
        }
        for (int i = rejected; i < candidates.length; i++) {
            if (i > rejected && candidates[i] == candidates[i - 1])
                continue;
            list.add(candidates[i]);
            backTrack(candidates, target - candidates[i], list, i + 1);
            list.remove(list.size() - 1);
        }
    }
}

File: Backtracking/combinationsum2.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        res = set()
        nums.sort()

        def dfs(i, curr, total):
            if (total == target):
                res.add(tuple(curr))
                return

            if (i >= len(nums) or total > target):
                return

            dfs(i + 1, curr, total)
            curr.append(nums[i])
            dfs(i + 1, curr, total + nums[i])
            curr.pop()
        
        dfs(0, [], 0)
        return [list(combination) for combination in res]